WordPress/gutenberg
# Interactivity API: Create `@wordpress/interactivity-router` module

## What?

Tracking issue: https://github.com/WordPress/gutenberg/issues/56803
Created on top of https://github.com/WordPress/gutenberg/pull/57859

Moves all the router-related code to its own module named `@wordpress/interactivity-router`.

The module exposes the `navigate` and `prefetch` functions previously exported from `@wordpress/interactivity`, now as actions of the `core/router` store.

As a reference, these are the bundle size results before and after splitting the code.

| asset | raw size (KiB) | gzipped size (KiB) |
| :-- | --: | --: |
| index.min.js (07a53b6a66c31eaba05d2597d0ab71281bd6a4c6) | 33.8 | 12.51 |
| index.min.js (144c15c3798dbb7bc37ae2dde4d8b5c33e1323a8) | 32.7 | 12.11 |
| router.min.js (144c15c3798dbb7bc37ae2dde4d8b5c33e1323a8) | 1.76 | 0.93 |

About the table:
1. The commit in _asset_ is the commit used to build the files.
2. Raw sizes are the values returned by Webpack during compilation.
3. Gzipped sizes are the values of the `Content-Length` header in HTTP responses.
4. Files are compiled for production.

## Why?

To start splitting the Interactivity API into separate parts.

## How?

- The `router.js` file is moved to its own folder and imports all the required dependencies from `@wordpress/interactivity` instead of using relative imports.
- A new entry point is created for that module.
- The `@wordpress/interactivity-router` module is registered in Gutenberg.
- The `@wordpress/interactivity-router` module is added as a dependency of the Query block's `view.js` module.

## Testing Instructions

1. In the site editor, go to the home template and ensure the Query block has "force page reload" disabled.
2. Visit the homepage.
3. Open the Network tab of the developer tools.
4. Ensure `interactivity/index.min.js` and `interactivity/query.min.js` have been requested, but not `interactivity/router.min.js`.
5. Hover the "Next Page" link.
6. Ensure `interactivity/router.min.js` has been requested.
7. Navigate to the next page.
8. Ensure the navigation happens without reloading the page.
